File: algorithm/src/baekjoon/java/bfs/BOJ1697.java
package baekjoon.java.bfs;

import java.io.*;
import java.util.*;

public class BOJ1697 {
    private static boolean[] line = new boolean[100001];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        if (n == k) {
            System.out.println(0);
            return;
        }
        Queue<Integer> queue = new LinkedList<>();
        queue.add(n);
        line[n] = true;

        int size;
        int count = 0;
        while (true) {
            count++;
            size = queue.size();
            for (int i = 0; i < size; i++) {
                int current = queue.poll();
                if (current-1 == k || current+1 == k || current*2 == k) {
                    System.out.println(count);
                    return;
                }
                int left = left(current);
                if (left >= 1 && !line[left]) {
                    line[left] = true;
                    queue.add(left);
                }
                int right = right(current);
                if (right <= 100000 && !line[right]) {
                    line[right] = true;
                    queue.add(right);
                }
                int teleport = teleport(current);
                if (teleport <= 100000 && !line[teleport]) {
                    line[teleport] = true;
                    queue.add(teleport);
                }
            }
        }

    }

    private static int left(int num) {
        return num - 1;
    }

    private static int right(int num) {
        return num + 1;
    }

    private static int teleport(int num) {
        return num * 2;
    }
}

File: algorithm/src/baekjoon/kotlin/boj/BOJ1697.kt
package baekjoon.kotlin.boj

import java.util.*

fun main() {
    val arr = readln().split(" ").map { it.toInt() }
    val line = BooleanArray(100_001)

    val queue: Queue<Int> = LinkedList()
    queue.add(arr[0])
    var answer = -1
    while (!queue.isEmpty()) {
        answer++
        val count = queue.size
        for (i in 1..count) {
            val poll = queue.poll()
            if (poll == arr[1]) {
                println(answer)
                return
            }

            line[poll] = true
            if (poll - 1 >= 0 && !line[poll - 1]) queue.add(poll - 1)
            if (poll + 1 <= 100_000 && poll + 1 <= arr[1] && !line[poll + 1]) queue.add(poll + 1)
            if (poll * 2 <= 100_000 && !line[poll * 2]) queue.add(poll * 2)
        }
    }
}
